### PR TITLE
Continuations: Don't re-run pipeline parts that can only be run once

### PR DESCRIPTION
**Pull request by georgwiese**
A typical continuations workflow looks like this:

```mermaid
 graph TD
     AsmFile --> ...
     ... --> PilWithConstants
     PilWithConstants --> GeneratedWitness
     GeneratedWitness --> Proof
    style GeneratedWitness fill:#ffe7e6
    style Proof fill:#ffe7e6
```

Only the two last steps are chunk-specific, yet we used to run the entire pipeline for each chunk.

With this PR, it is advanced to the `PilWithConstants` stage once, the result of which is kept in memory. This artifact is now clonable (rather than the entire pipeline as suggested in #840). As a result, the pipeline can be re-started from this step.

### CODE REVIEW

**Comment by georgwiese** [comment]
Needs to be renamed after #869 is merged.

File: riscv/src/continuations.rs [hunk 1]
```diff
@@ -45,15 +45,23 @@ where
...
 
+    log::info!("Advancing pipeline to PilWithEvaluatedFixedCols stage...");
+    let pipeline = pipeline_factory();
+    let pil_with_evaluated_fixed_cols = pipeline.pil_with_evaluated_fixed_cols().unwrap();
+
+    // This returns the same pipeline as pipeline_factory() (with the same name, output dir, etc...)
+    // but starting from the PilWithEvaluatedFixedCols stage. This is more efficient, because we can advance
+    // to that stage once before we branch into different chunks.
+    let optimized_pipeline_factory = || {
+        pipeline_factory().from_pil_with_evaluated_fixed_cols(pil_with_evaluated_fixed_cols.clone())
```
Comment (chriseth) [inline riscv/src/continuations.rs:56]: Instead of cloning, you could also change `PilWithEvaluatedFixedCols` to contain an `Rc` to the values. Same with evaluated fixed columns, they are never modified once they were created, so`Rc` seems to be perfectly suitable.

Comment (georgwiese) [reply riscv/src/continuations.rs:56]: Like this? d1a1bb2f0e0741bf46a4d7c5cac412eb777f9299

File: ast/src/analyzed/mod.rs [hunk 2]
```diff
@@ -23,7 +23,7 @@ pub enum StatementIdentifier {
     Identity(usize),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:26]: Now do we still need this? :)
We can keep it if you want.

File: compiler/benches/executor_benchmark.rs [hunk 3]
```diff
@@ -26,16 +26,11 @@ fn criterion_benchmark(c: &mut Criterion) {
     let contents = compiler::compile(riscv_asm_files, &CoProcessors::base(), false);
     let pil_with_constants = Pipeline::<T>::default()
         .from_asm_string(contents, None)
-        .optimized_pil_with_constants()
+        .pil_with_evaluated_fixed_cols()
```
Comment (chriseth) [inline compiler/benches/executor_benchmark.rs:29]: Is this a rebase artifact?

File: compiler/src/pipeline.rs [hunk 4]
```diff
@@ -158,7 +160,7 @@ where
             log_level: Level::Debug,
             name: None,
             force_overwrite: false,
-            tmp_dir: None,
+            _tmp_dir: None,
```
Comment (chriseth) [inline compiler/src/pipeline.rs:163]: Why did you add the underscore?

File: ast/src/analyzed/mod.rs (hunk 2 above)
Comment (georgwiese) [reply ast/src/analyzed/mod.rs:26]: Nope, will remove :)

File: compiler/benches/executor_benchmark.rs (hunk 3 above)
Comment (georgwiese) [reply compiler/benches/executor_benchmark.rs:29]: No, I renamed the function in this PR to be more consistent with the naming introduced in yours! It returns a `PilWithEvaluatedFixedCols`.

File: compiler/src/pipeline.rs (hunk 4 above)
Comment (georgwiese) [reply compiler/src/pipeline.rs:163]: Because it is no longer used (and the linter complains about it unless you start with underscore). As explained in the docstring of `_tmp_dir`, it is only kept in Pipeline so that it has an owner. The directory will live for as long as the pipeline does.

**Comment by georgwiese** [comment]
Fixed the unnecessarily derived `Clone` and squashed.

**Review by leonardoalt — DISMISSED** [review]

**Comment by georgwiese** [comment]
Rebased onto main

**Review by leonardoalt — APPROVED** [review]
